File: problems/2/main.js
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next);
}

/**
 * @param {ListNode} node
 */
var printNumber = function (node) {
    while (node != null) {
        process.stdout.write(node.val.toString());
        if (node.next != null) {
            process.stdout.write(" -> \0");
        }
        node = node.next;
    }
    console.log();
}

/**
 * @param {ListNode} node
 * @return {ListNode}
 */
var goNext = function (node) {
    node.next = new ListNode();
    return node.next;
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    const PTR = new ListNode();
    var current = PTR,
        sum, carry = 0;

    while (l1 != null || l2 != null || carry) {
        sum = carry;

        if (l1 != null) {
            sum += l1.val;
            l1 = l1.next;
        }

        if (l2 != null) {
            sum += l2.val;
            l2 = l2.next;
        }

        carry = (sum > 9) ? 1 : 0;
        current = goNext(current);
        current.val = sum % 10;
    }

    return PTR.next;
};

(function main() {
    const a = [9, 9, 9, 9, 9, 9, 9],
        b = [9, 9, 9, 9];

    l1 = new ListNode();
    temp = l1;
    for (let x in a) {
        temp.val = a[x];
        if (x != a.length - 1)
            temp = goNext(temp);
    }

    l2 = new ListNode();
    temp = l2;
    for (let x in b) {
        temp.val = b[x];
        if (x != b.length - 1)
            temp = goNext(temp);
    }

    result = addTwoNumbers(l1, l2);
    printNumber(l1);
    printNumber(l2);
    console.log("--------------------------");
    printNumber(result);
})();

File: problems/2/main.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
  int val;
  struct ListNode *next;
};

void printNumber(struct ListNode *node) {
  while (node != NULL) {
    printf("%d", node->val);
    if (node->next != NULL) {
      printf(" -> ");
    }
    node = node->next;
  }
  printf("\n");
}

/*
    Creates a new node to be the next of the given one
    and make the pointer point to it
 */
void goNext(struct ListNode **node) {
  (*node)->next = (struct ListNode *)malloc(sizeof(struct ListNode));
  *node = (*node)->next;
  (*node)->next = NULL;
}

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
  // PTR
  struct ListNode *PTR = (struct ListNode *)malloc(sizeof(struct ListNode));
  struct ListNode *current = PTR;
  int sum, carry = 0;

  // Move between places in the number
  while (l1 != NULL || l2 != NULL || carry) {
    sum = carry;

    if (l1 != NULL) {
      sum += l1->val;
      l1 = l1->next;
    }

    if (l2 != NULL) {
      sum += l2->val;
      l2 = l2->next;
    }

    carry = sum > 9;
    goNext(&current);
    current->val = sum % 10;
  }

  return PTR->next;
}

int main() {

  struct ListNode *l1, *l2, *temp, *result;

  int a[] = {9, 9, 9, 9, 9, 9, 9}, /**/
      b[] = {9, 9, 9, 9};

  size_t size;

  // First number
  // [2, 4, 3]
  // PTR: 2
  l1 = (struct ListNode *)malloc(sizeof(struct ListNode));
  temp = l1;
  size = sizeof(a) / sizeof(a[0]);
  for (int i = 0; i < size; i++) {
    temp->val = a[i];
    if (i != size - 1)
      goNext(&temp);
  }

  // Second number
  // [5, 6, 4]
  // PTR: 5
  l2 = (struct ListNode *)malloc(sizeof(struct ListNode));
  temp = l2;
  size = sizeof(b) / sizeof(b[0]);
  for (int i = 0; i < size; i++) {
    temp->val = b[i];
    if (i != size - 1)
      goNext(&temp);
  }

  result = addTwoNumbers(l1, l2);
  printNumber(l1);
  printNumber(l2);
  printf("--------------------------\n");
  printNumber(result);

  return 0;
}